src/lib.rs

```diff
@@ -11,6 +11,678 @@
+            "mqtts"
+        } else {
+            "mqtt"
+        };
+
+        let mqtt_uri = format!(
+            "{}://{}:{}",
+            mqtt_protocol, config.mqtt_hostname, config.mqtt_port
+        );
+
+        let mqtt_cli = mqtt::CreateOptionsBuilder::new()
+            .server_uri(mqtt_uri)
+            .client_id(client_id)
+            .max_buffered_messages(100)
+            .create_client()
+            .map_err(|e| {
+                UStatus::fail_with_code(
+                    UCode::INTERNAL,
+                    format!("Unable to create mqtt client: {e:?}"),
+                )
+            })?;
+
+        mqtt_cli.set_message_callback(
+            move |cli: &AsyncClient, message: std::option::Option<paho_mqtt::Message>| {
+                UPClientMqtt::on_receive(cli, message, ptr.clone())
+            },
+        );
+
+        // TODO: Integrate ssl options when connecting, may need a username, etc.
+        let conn_opts = mqtt::ConnectOptionsBuilder::with_mqtt_version(MQTT_VERSION_5)
+            .clean_start(false)
+            .properties(mqtt::properties![mqtt::PropertyCode::SessionExpiryInterval => 3600])
+            .finalize();
+
+        mqtt_cli.connect(conn_opts).await.map_err(|e| {
+            UStatus::fail_with_code(
+                UCode::INTERNAL,
+                format!("Unable to connect to mqtt broker: {e:?}"),
+            )
+        })?;
+
+        Ok(Self {
+            mqtt_client: Arc::new(mqtt_cli),
+            topic_listener_map,
+            authority_name,
+            client_type,
+        })
+    }
+
+    /// Helper function that handles MQTT messages on reception.
+    ///
+    /// # Arguments
+    /// * `cli` - MQTT client that received the message.
+    /// * `message` - MQTT message received.
+    fn on_receive(
+        _cli: &AsyncClient,
+        message: Option<paho_mqtt::Message>,
+        topic_map: Arc<RwLock<HashMap<String, HashSet<ComparableListener>>>>,
+    ) {
+        if let Some(msg) = message {
+            let topic = msg.topic();
+
+            // Get attributes from mqtt header.
+            let uattributes =
+                UPClientMqtt::get_uattributes_from_mqtt_properties(msg.properties()).unwrap();
+
+            let payload = msg.payload();
+            let upayload = payload.to_vec();
+
+            let umessage = UMessage {
+                attributes: Some(uattributes).into(),
+                payload: Some(upayload.into()),
+                ..Default::default()
+            };
+
+            // Create UMessage from UAttributes and UPayload.
+            let topic_map_read = block_on(topic_map.read());
+            let listeners = topic_map_read.get(topic);
+
+            if let Some(listeners) = listeners {
+                for listener in listeners {
+                    let umsg_clone = umessage.clone(); // need to clone outside of closure.
+                    block_on(listener.on_receive(umsg_clone));
```

Comment:
Do we need block_on here? Can't we .await?

Reply:
See other comment about set_message_callback not accepting an async function